src/txn/txn_ckpt.c

```diff
@@ -705,46 +751,6 @@ __checkpoint_prepare(
 	txn_state->id = txn_state->pinned_id =
 	    txn_state->metadata_pinned = WT_TXN_NONE;
 
-#ifdef HAVE_TIMESTAMPS
-	/*
-	 * Set the checkpoint transaction's timestamp, if requested.
-	 *
-	 * We rely on having the global transaction data locked so the oldest
-	 * timestamp can't move past the stable timestamp.
-	 */
-	WT_ASSERT(session, !F_ISSET(txn,
-	    WT_TXN_HAS_TS_COMMIT | WT_TXN_HAS_TS_READ |
-	    WT_TXN_PUBLIC_TS_COMMIT | WT_TXN_PUBLIC_TS_READ));
-
-	if (use_timestamp) {
-		/*
-		 * If the user wants timestamps then set the metadata
-		 * checkpoint timestamp based on whether or not a stable
-		 * timestamp is actually in use.  Only set it when we're not
-		 * running recovery because recovery doesn't set the recovery
-		 * timestamp until its checkpoint is complete.
-		 */
-		if (txn_global->has_stable_timestamp) {
-			__wt_timestamp_set(&txn->read_timestamp,
-			    &txn_global->stable_timestamp);
-			F_SET(txn, WT_TXN_HAS_TS_READ);
```

Comment:
The bug is caused by this code that sets a read timestamp on a running transaction.